rust/azure_iot_operations_services/examples/azure_device_registry_client.rs

```diff
@@ -0,0 +1,344 @@
+        application_context,
+        session.create_managed_client(),
+        azure_device_registry::ClientOptions::default(),
+    )?;
+
+    // Run the Session and the Azure Device Registry operations concurrently
+    let r = tokio::join!(
+        azure_device_registry_operations(
+            azure_device_registry_client,
+            session.create_exit_handle()
+        ),
+        session.run(),
+    );
+    r.1?;
+    Ok(())
+}
+
+async fn azure_device_registry_operations(
+    azure_device_registry_client: azure_device_registry::Client<SessionManagedClient>,
+    exit_handle: SessionExitHandle,
+) {
+    let timeout = Duration::from_secs(10);
+
+    // observe for updates for our Device + Inbound Endpoint
+    // TODO: uncomment once service supports this
+    // match azure_device_registry_client
+    //     .observe_device_update_notifications(
+    //         DEVICE_NAME.to_string(),
+    //         INBOUND_ENDPOINT_NAME.to_string(),
+    //         timeout,
+    //     )
+    //     .await
+    // {
+    //     Ok(mut observation) => {
+    //         log::info!("Device observed successfully");
+    //         tokio::task::spawn({
+    //             async move {
+    //                 while let Some((notification, _)) = observation.recv_notification().await {
+    //                     log::info!("device updated: {notification:#?}");
+    //                 }
+    //                 log::info!("device notification receiver closed");
+    //             }
+    //         });
+    //     }
+    //     Err(e) => {
+    //         log::error!("Observing for device updates failed: {e}");
+    //     }
+    // };
+
+    // // observe for updates for our Asset
+    // TODO: uncomment once service supports this
+    // match azure_device_registry_client
+    //     .observe_asset_update_notifications(
+    //         DEVICE_NAME.to_string(),
+    //         INBOUND_ENDPOINT_NAME.to_string(),
+    //         ASSET_NAME.to_string(),
+    //         timeout,
+    //     )
+    //     .await
+    // {
+    //     Ok(mut observation) => {
+    //         log::info!("Asset observed successfully");
+    //         tokio::task::spawn({
+    //             async move {
+    //                 while let Some((notification, _)) = observation.recv_notification().await {
+    //                     log::info!("asset updated: {notification:#?}");
+    //                 }
+    //                 log::info!("asset notification receiver closed");
+    //             }
+    //         });
+    //     }
+    //     Err(e) => {
+    //         log::error!("Observing for asset updates failed: {e}");
+    //     }
+    // };
+
+    // Get Device + Inbound Endpoint details and send status update
+    match azure_device_registry_client
+        .get_device(
+            DEVICE_NAME.to_string(),
+            INBOUND_ENDPOINT_NAME.to_string(),
+            timeout,
+        )
+        .await
+    {
+        Ok(device) => {
```

Comment:
How scalable do we think this pattern is? If customers end up copying and pasting this, and trying to extend it, is this scoping going to become an issue? Could we propagate error instead to reduce the conditions/scopes?

Reply:
updated per our discussion to have device and asset operations run in separate functions so that they can use `?` instead of nesting